crates/conjure_core/src/rules/minion.rs

```diff
@@ -221,10 +214,8 @@ fn flatten_eq(expr: &Expr, model: &Model) -> ApplicationResult {
     }
 
     let expr = expr.with_children(children);
-    let new_top = exprs_to_conjunction(&new_tops)
-        .unwrap_or_else(|| bug!("rules::minion::flatten_eq: new_tops could be combined with And!"));
 
-    Ok(Reduction::new(expr, new_top, model.variables))
```

Comment:
@niklasdewally I just notised that in flatten binop I just put new_tops in our reduction. However, I noticed that you used to check if all expressions in new_tops are booleans, and if not - you threw a bug!.

Should I do the same? I wrote this, but I'm not sure if it's necessary 

```rust
/// Checks if all expressions in `Vec<Expr>` are booleans. This needs to be true so
/// the vector could be conjuncted to the model.
/// # Returns
///
/// - true: the expression list is non empty, and all expressions are booleans (so can be
/// conjuncted).
///
/// - false: the expression list is empty, or not all expressions are booleans.
pub fn is_vec_bool(exprs: &Vec<Expr>) -> bool {
    match exprs.as_slice() {
        [] => false,
        [a] if a.return_type() == Some(ReturnType::Bool) => true,
        _ => {
            for expr in exprs {
                if expr.return_type() != Some(ReturnType::Bool) {
                    return false;
                }
            }
            true
        }
    }
}
```

```rust
    if is_vec_bool(&new_tops) {
        Ok(Reduction::new(expr, new_tops, model.variables))
    } else {
        bug!("rules::minion::flatten_eq: new_tops could be combined with And!");
    }
```

Reply:
> However, I noticed that you used to check if all expressions in new_tops are booleans, and if not - you threw a bug!.

To be correct type, the top level should all be bools, that's correct. I now think that this check might be better within the rewriter instead of specific rules, as this is true for all rules, not just `flatten_binop`. I.e. add a debug assert after each rule application in the rewriter, checking that the top level constraints are all still bools. 

Not necessary by any means, but being defensive here could help us identify broken rules faster.

Reply:
Ideally, the rewriter should do as much rule-agnostic checking work as possible, instead of trusting the rules to do things properly; however, that is definitely outside the scope of this PR.